include/rocksdb/utilities/object_registry.h

```diff
@@ -11,80 +11,193 @@
+  class Library {
+  private:
+    friend ObjectRegistry;
+    
+    // ** FactoryFunctions for this loader, organized by type
+    std::unordered_map<std::string,
+		       std::vector<std::unique_ptr<Entry> > > entries_;
+    // Finds the entry matching the input name and type
+    const Entry* FindEntry(const std::string& type,
+			   const std::string& name) const;
+    void AddEntry(const std::string & type, std::unique_ptr<Entry> & entry);
+  public:
+    void Dump(Logger *logger) const;
+    // To register a factory function for a type T, initialize a Registrar<T>
+    // object with static storage duration. For example:
+    template <typename T>
+    const FactoryFunc<T>& Register(const std::string& pattern,
+                                 const FactoryFunc<T>& factory) {
+      std::unique_ptr<Entry> entry(new FactoryEntry<T>(pattern, factory));
+      AddEntry(T::Type(), entry);
+      return factory;
+    }
+  };
+  
+private:
+  std::vector<std::shared_ptr<Library>> libraries_;
  public:
-  explicit Registrar(std::string pattern, FactoryFunc<T> factory);
-};
+  static std::shared_ptr<Library>& Default();
+  static std::shared_ptr<ObjectRegistry> Get();
 
-// Implementation details follow.
-
-namespace internal {
-
-template <typename T>
-struct RegistryEntry {
-  std::regex pattern;
-  FactoryFunc<T> factory;
-};
+  ObjectRegistry();
 
-template <typename T>
-struct Registry {
-  static Registry* Get() {
-    static Registry<T> instance;
-    return &instance;
+  std::shared_ptr<Library> AddLibrary() {
+    std::shared_ptr<Library> library = std::make_shared<Library>();
+    libraries_.emplace_back(library);
+    return library;
   }
-  std::vector<RegistryEntry<T>> entries;
 
- private:
-  Registry() = default;
-};
+  // Creates a new T using the factory function that was registered with a
+  // pattern that matches the provided "target" string according to
+  // std::regex_match.
+  //
+  // If no registered functions match, returns nullptr. If multiple functions
+  // match, the factory function used is unspecified.
+  //
+  // Populates res_guard with result pointer if caller is granted ownership.
+  template <typename T>
+  T* NewObject(const std::string& target, std::unique_ptr<T>* guard,
+		     std::string *errmsg) {
+    guard->reset();
+    const auto* basic = FindEntry(T::Type(), target);
+    if (basic != nullptr) {
+      const auto* factory = static_cast<const FactoryEntry<T>*>(basic);
+      return factory->NewFactoryObject(target, guard, errmsg);
+    } else {
+      *errmsg = std::string("Could not load ") + T::Type();
+      return nullptr;
+    }
+  }
 
-}  // namespace internal
+  // Creates a new shared T using the input factory functions.
+  // Returns OK if a new shared T was successfully created
+  // Returns NotFound if the type/target could not be created
+  // Returns InvalidArgument if the factory return an unguarded object
+  //                      (meaning it cannot be managed by a shared ptr)
```

Comment:
shared ptr or unique_ptr?

Reply:
I think there is a typo with the comment. You are trying to obtain a unique_ptr to the object, thus I expect the comment means: if the object cannot be managed by a unique_ptr, then it should return InvalidArgument. However the original text in the comment says 'shared ptr'.